src/TopoMojo.Api/Features/User/ClaimsTransformation.cs

```diff
@@ -61,6 +48,116 @@ public async Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
 
+        /// <summary>
+        /// Determines if the current ClaimsPrincipal contains a claim that indicates that it is attempting to auth as a
+        /// service account (via OAuth Client Credentials).
+        /// </summary>
+        /// <param name="principal"></param>
+        /// <returns>The clientID of the OAuth client eligible for auth if configured correctly, or null if not.</returns>
+        private string ResolveServiceAccountClientId(ClaimsPrincipal principal)
+        {
+            var serviceAccountClientId = string.Empty;
+
+            // if authenticating as a service account, appropriate OIDC options (AuthTypeClaimName and ServiceAccountAuthType) must be configured
+            // and the token must have the appropriate claim and value
+            if (oidcOptions.AuthTypeClaimName.IsEmpty())
+            {
+                logger.LogInformation("Auth type claim not configured.");
+                return null;
+            }
+
+            var claim = principal.Claims.FirstOrDefault(c => c.Type == oidcOptions.AuthTypeClaimName);
+            if (claim is null)
+            {
+                logger.LogInformation("Auth type claim {claimName} not present.", oidcOptions.AuthTypeClaimName);
+                return null;
+            }
+
+            if (claim.Value != oidcOptions.ServiceAccountAuthType)
+            {
+                logger.LogInformation("Auth type claim {claimName} has unexpected value {claimValue}", claim.Type, claim.Value);
+                return null;
+            }
+
+            var clientIdClaim = principal.Claims.FirstOrDefault(c => c.Type == "client_id" || c.Type == "azp");
+            if (clientIdClaim is null)
+            {
+                logger.LogInformation("Couldn't resolve a client ID from the ClaimsPrincipal.");
+                return null;
+            }
+
+            return clientIdClaim.Value;
+        }
+
+        /// <summary>
+        /// This transformation supports two paths to resolution of a user:
+        ///     1. A typical subject-claim based user identification
+        ///     2. A claim that identifies the token as belong to a service account which matches an existing user's ServiceAccountClientId property.
+        /// 
+        /// The claim identifying service accounts will be consulted first if it is configured in Topo's OIDC settings and present in the claims principal. Otherwise,
+        /// this function falls back to the subject claim to match to a user.
+        /// </summary>
+        /// <param name="principal"></param>
+        /// <returns>A Topomojo User representing the result of the resolution process.</returns>
+        private async Task<User> ResolveUser(ClaimsPrincipal principal)
+        {
+            var subject = principal.Subject();
+            var serviceAccountClientId = ResolveServiceAccountClientId(principal);
+            var resolvedUser = default(User);
+
+            if (subject.IsEmpty() && !serviceAccountClientId.IsEmpty())
```

Comment:
Is the ! supposed to be here? It's checking if both subject and serviceAccountClientId are empty right?